### PR TITLE
build-tools: Add new tasks for build caching

### PR DESCRIPTION
**Pull request by tylerbutler**
Makes several updates to fluid-build so that it caches results better.

- Updates the api-extractor task to include additional config files in its cache key. This does mean that all api-extractor tasks are invalidated if any of the config files change. However, it still doesn't take into account the shared configs that the leaf configs inherit from. So changes to the root config do not invalidate the cache. Adding that is a larger change.
- Adds an option to the LeafWithFileStatDoneFileTask that enables the DoneFile it creates to use file contents hashes instead of file stats. This is useful for tasks whose metadata may be legitimately updated during the build. This is not currently used, but it will likely be needed for some upcoming tasks.
- Adds a new flub command, `rename-types`, that will replace renamer because renamer complains loudly when there are no source files, and it needs to run without issue in that scenario.

### CODE REVIEW

File: build-tools/packages/build-tools/src/fluidBuild/tasks/taskFactory.ts [hunk 1]
```diff
@@ -46,6 +47,11 @@ const executableToLeafTask: {
 	"flub generate typetests": TypeValidationTask,
 	"fluid-type-test-generator": TypeValidationTask,
 	"depcruise": DepCruiseTask,
+
+	// Note that this assumes that "renamer" is ONLY used for renaming types. If it is used in a different task in the
+	// pipeline then this mapping will have to be updated.
+	"renamer": RenameTypesTask,
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/taskFactory.ts:53]: I added a mapping for renamer to ease migration.

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts [hunk 2]
```diff
@@ -565,4 +572,31 @@ export abstract class LeafWithFileStatDoneFileTask extends LeafWithDoneFileTask
 			return undefined;
 		}
 	}
+
+	private async getHashDoneFileContent(): Promise<string | undefined> {
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:576]: I'm partial to not adding this now and keep it in a branch, but no strong feelings about it.

**Review by alexvy86 — APPROVED** [review]
Seems reasonable based on my understanding of the build tasks system.

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts (hunk 2 above)
Comment (jason-ha) [reply build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/leafTask.ts:576]: It definitely needs more comments.
I have questions about the whole system. Hopefully I can find enough context in some other files (that aren't changed here).

File: build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/renamerTask.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * This task enables caching of the results of renaming ESM types files, which we do in our build pipeline to work
+ * around a limitation in tsc-multi.
+ *
+ * This implementation is deliberately unintelligent. It assumes that .d.ts files are renamed to .d.mts in the lib
+ * folder. Since renaming files doesn't leave the source files around, this implementation fakes source files by
+ * treating the .d.ts files in the dist folder as the input and the renamed files in the lib folder as the output. This
+ * is safe because any time the output in the dist folder is updated then it's very likely the lib folder would also
+ * need to be updated for other reasons. In other words this shouldn't cause a lot of unnecessary cache invalidation.
```
Comment (jason-ha) [inline build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/renamerTask.ts:16]: Can we add a comment that this is not sufficient to trigger every time running the task is warranted?

Seems like a better fake would be to use all of the `lib/**/*.mts` files excluding `*.d.mts`.
Alternatively, add smarts to look for files that shouldn't exist and if they do, then trigger task. This task is ideally triggered at least every time a `lib/**/*.d.ts ` exists.

I am also wondering how this system (for rename and overall) deals with files that are deleted. They just sit around and hope they cause no harm?

Comment (jason-ha) [reply build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/renamerTask.ts:16]: Proposed fix #18896

**Pull request merged**